quodlibet/quodlibet/qltk/paned.py

```diff
@@ -119,6 +120,210 @@ class RVPaned(RPaned):
     ORIENTATION = Gtk.Orientation.VERTICAL
 
 
+class XPaned(Paned):
+    """A Paned with addition support for expander widget children."""
+
+    ORIENTATION = None
+
+    def __init__(self, *args, **kwargs):
+        if self.ORIENTATION is not None:
+            kwargs["orientation"] = self.ORIENTATION
+        super(XPaned, self).__init__(*args, **kwargs)
+
+        self.connect('button-release-event', self.__button_release)
+        self.connect('check_resize', self.__check_resize)
+
+        self.__disablenotify_id = None
+
+        self.__panelocks = []
+        self.__updating = False
+        self.__resizing = False
+
+    def __button_release(self, *data):
+        for w in self.__panelocks:
+            alloc = w.get_size_request()
+            size = 0
+            if self.ORIENTATION == Gtk.Orientation.VERTICAL:
+                size = alloc[1]
+                w.set_size_request(-1, size)
+            else:
+                size = alloc[0]
+                w.set_size_request(size, -1)
+#            print_d("%r | resetting locked size %d" % (w.id, size))
+        del self.__panelocks
+        self.__panelocks = []
+
+    def panelocks(self, widget, has=True):
+        widgets = []
+        if isinstance(widget.get_child1(), XPaned):
+            children = self.panelocks(widget.get_child1(), has)
+            if children:
+                widgets.extend(children)
+        elif widget.get_child1() and \
+                 hasattr(widget.get_child1(), "panelock") == has:
+            widgets.append(widget.get_child1())
+        if isinstance(widget.get_child2(), XPaned):
+            children = self.panelocks(widget.get_child2(), has)
+            if children:
+                widgets.extend(children)
+        elif widget.get_child2() and \
+                 hasattr(widget.get_child2(), "panelock") == has:
+            widgets.append(widget.get_child2())
+        return widgets
+
+    def update(self, widget):
+        """Pass the widget whose pane needs updating."""
+
+        self.__updating = True
+
+        if not isinstance(widget, Gtk.Expander):
+            return
+
+        # lock other non-expander or expanded widgets
+        widgets = self.panelocks(self.root)
+        for w in (w for w in widgets
+            if w is not widget
+                 and (not isinstance(w, Gtk.Expander) or w.get_expanded())):
```

Comment:
So long as we (this class) controls all the packing, I think we should be able to omit the various instance checks, they make me OOP-nervous anyway

Reply:
the trouble is, we need to consider other non-expander types that are in the mix too! again, 'main' (now a ConfigRHPaned obj) and 'playbar' (TopBar obj) are parts of resizing equation. it did look nice and clean before i started working with two or three nested paneds, and that's when all heck broke loose. that chunk of code is now very simple to look though.